### ISSUE TITLE
False positive on no-use-before-define with Promise.then(x => y) syntax

### ISSUE THREAD

**Issue by wanaryytel** [eslint/eslint#12645]
**Tell us about your environment**

Environment Info:

Node version: v10.16.0
npm version: v6.9.0
Local ESLint version: v6.7.1 (Currently used)
Global ESLint version: Not found

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
    "parser": "babel-eslint",
    "extends": ["airbnb"],
    "env": {
        "browser": true,
    },
    "globals": {
        "FCONST": false,
        "CONSTANTS": false,
        "URLS": false,
    },
    "settings": {
        "import/resolver": {
            "webpack": {
                "config": "webpack/config.dev.js",
            },
        },
    },

    "plugins": [
        "react-hooks"
    ],

    "rules": {
        // We use 4 spaces for indentation. Everywhere.
        "indent": ["error", 4, {"SwitchCase": 1}],
        "react/jsx-indent": ["error", 4],
        "react/jsx-indent-props": ["error", 4],

        // JSX files use .js extension
        "react/jsx-filename-extension": ["error", {"extensions": [".js"]}],

        // Pure components are more efficient in some cases, but written as class components
        "react/prefer-stateless-function": ["error", {"ignorePureComponents": true}],

        "react-hooks/rules-of-hooks": "error",
        "react-hooks/exhaustive-deps": "warn",

        // Set max line length to 120 chars. Smaller lengths feel too small on modern monitors.
        "max-len": ["error", 120],

        // Allow only single quotes.
        "quotes": ["error", "single"],

        // else-return improves readability sometimes, especially with one-liners.
        "no-else-return": "off",

        // Don't keep spaces inside of curly braces for objects.
        "object-curly-spacing": ["error", "never"],

        "curly": ["error", "all"],

        // Allow unary ++ operator in for loop afterthoughts
        "no-plusplus": ["error", {"allowForLoopAfterthoughts": true}],

        // allow some underscore dangles
        "no-underscore-dangle": ["error"],

        // Turn these into warnings
        "class-methods-use-this": "warn",
        "jsx-a11y/no-static-element-interactions": "warn",

        // Show TODOs and FIXMEs as warnings
        "no-warning-comments": ["warn", {"location": "anywhere"}],

        // Initialise state however you want
        "react/state-in-constructor": "off"
    },
    "overrides": [
        {
            "files": ["store.js"],
            "rules": {
                "no-param-reassign": ["error", { "props": true, "ignorePropertyModificationsFor": ["state"] }]
            }
        }
    ]
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
// Short version:

const bar = useBar({
    foo,
});

const foo = () => {
    axios.get(url)
        .then(() => {
            bar.set('x');
        });
};



// Longer more logical version
const onSubmit = (values) => {
    axios.post(url, values)
        .then(() => {
            formik.setSubmitting(false);
            onFinish();
        });
};

const formik = useFormik({
    validationSchema: MySchema,
    initialValues: {
        users: [],
    },
    onSubmit,
});
```

```bash
eslint .
```

**What did you expect to happen?**
It's a no-win situation right now because there is no good way to fix this issue. The `no-use-before-define` rule will be true no matter which way you order things even though the `onSubmit` (or `foo`) should obviously come first. Right now the linter doesn't detect that it's called in the Promise deferred `then` method.

**What actually happened? Please include the actual, raw output from ESLint.**

```
error    'foo' was used before it was defined     no-use-before-define

// or

error    'formik' was used before it was defined  no-use-before-define
```
**Are you willing to submit a pull request to fix this bug?**
Not right now, maybe in the future

**Comment by kaicataldo** [comment]
Thanks for the detailed information! Are you able to share the configuration for `no-use-before-define`, just so we have all the information? You can use the [`--print-config`](https://eslint.org/docs/user-guide/command-line-interface#print-config) CLI option on one of your source files to get that info.

**Comment by wanaryytel** [comment]
Sorry, I didn't notice this for a while. Of course, here we go:

```js
"no-use-before-define": [
      "error",
      {
        "functions": true,
        "classes": true,
        "variables": true
      }
    ],

```

**Comment by ljharb** [comment]
This seems like a correct warning to me anyways - "before" doesn't refer to time, it refers to lexical position in the file.
